62-california-condor/technical_specification.md

```diff
@@ -0,0 +1,192 @@
+schedule updates for low-traffic times and even notify users beforehand if necessary.
+If we determine that migrations take long enough that the above approach becomes
+unrealistic, we'll have to take a more complicated approach.
+
+
+### Reverse Migrations
+
+There might be some situation where we need to apply the reverse of a migration.
+If DB version 5 is applied, but we later find that we should have stayed with version 4,
+then we need to move to version 6. Version 6 is not treated as some special 'undo'
+version increment; the migration logic merely happens to move the data to the same
+structure contained in version 4. However, in the case that we really just need to
+deploy an older service version, the migration code needs to be able to perform
+migrations from X to X - 1. The responsibility falls to the developer to ensure their
+forward migration code has a backward counterpart, which should also be tested.
+
+### Migration Code Organization
+
+![Migration code organization](./images/migration%20code%20org.png)
+
+In the image above, the green section indicates the top-level migration logic, most of
+which can be implemented in a library like `hexkit`.  
+The *rounded* orange boxes represent distinct migrations between database versions.  
+The *square* orange boxes show common logic that can be abstracted into a library.  
+The red-dotted items would be performed once for each batch of documents if a collection
+were to be processed in batches.  
+The gray box represents custom migration logic.  
+This layout allows for simpler migrations to use migration framework code, but also
+enables us to write the code we need for more complex migrations, like merging two
+collections. The number of collections for any service is minimal, so there is no
+need to automate or enforce iteration through collections. In fact, that would hinder
+our ability to write migration logic involving multiple collections at once.
+
+### Errors During a Migration
+
+We can prevent database corruption from errors during migrations through the use of
+temp tables and some renaming. Here's an example of how that looks while migrating a database containing two collections named `users` and `orders`:
+1. Drop `tmp_users` and `tmp_orders` if they exist (maybe from previous failed attempt).
+2. Create `tmp_users` and `tmp_orders`.
+3. Read data from `users` and write new data to `tmp_users`.
+4. Read data from `orders` and write new data to `tmp_orders`.
+5. Rename `users` to `old_users` and `orders` to `old_orders`.
+6. Rename `tmp_users` to `users` and `tmp_orders` to `orders`.
+7. Apply the indexes from `old_users` and `old_orders` to `users` and `orders`.
+8. Do some verification (count docs, examine content, double check indexes, etc.).
+9. Drop `old_users` and `old_orders`.
+
+If an error occurs at any point, drop all `tmp_` tables from this migration, log the
+error, unset the lock document, make sure the old tables have the original names, and
+exit.
+
```

Comment:
Regarding prefixes, maybe `tmp_new_` instead of `tmp_` and `tmp_old_` instead of `_old` ?